File: java/src/main/java/org/endeavourhealth/library/model/iml/Concept.java
package org.endeavourhealth.library.model.iml;

import lombok.Getter;
import lombok.Setter;
import org.endeavourhealth.library.model.tripletree.TTIriRef;

import java.util.HashSet;
import java.util.Set;

@Getter
public class Concept extends Entity {

  private String code;
  private String im1Id;
  private Set<TTIriRef> subClassOf;
  private Set<Concept> matchedFrom;
  private Integer usage;
  private String codeId;
  private String alternativeCode;
  @Getter
  @Setter
  private boolean subsumed;


  public Concept setAlternativeCode(String alternativeCode) {
    this.alternativeCode = alternativeCode;
    return this;
  }

  public Concept setCodeId(String codeId) {
    this.codeId = codeId;
    return this;
  }

  public Concept setSubClassOf(Set<TTIriRef> subClassOf) {
    this.subClassOf = subClassOf;
    return this;
  }

  public Concept addSubClassOf(TTIriRef superClass) {
    if (this.subClassOf == null)
      this.subClassOf = new HashSet<>();
    this.subClassOf.add(superClass);
    return this;
  }

  public Concept setIri(String iri) {
    super.setIri(iri);
    return this;
  }

  public Concept setUsage(Integer usage) {
    this.usage = usage;
    return this;
  }


  public Concept setName(String name) {
    super.setName(name);
    return this;
  }

  public Concept setDescription(String description) {
    super.setDescription(description);
    return this;
  }

  public Concept setCode(String code) {
    this.code = code;
    return this;
  }


  public Concept setScheme(TTIriRef scheme) {
    super.setScheme(scheme);
    return this;
  }

  public Concept setMatchedFrom(Set<Concept> matchedFrom) {
    this.matchedFrom = matchedFrom;
    return this;
  }

  public Concept addMatchedFrom(Concept legacy) {
    if (this.matchedFrom == null)
      this.matchedFrom = new HashSet<>();
    this.matchedFrom.add(legacy);
    return this;
  }

  public void setIm1Id(String im1Id) {
    this.im1Id = im1Id;
  }
}

File: src/models/Concept.ts
import z from "zod";

import { Entity, EntitySchema } from "./Entity";
import { TTIriRef, TTIriRefSchema } from "./TTIriRef";

// export interface Concept extends Entity {
//   subClassOf?: TTIriRef[];
//   code?: string;
//   im1Id?: string;
//   matchedFrom?: Concept[];
//   usage?: number;
//   codeId?: string;
//   alternativeCode?: string;
//   subsumed?: boolean;
// }

export const ConceptSchema = EntitySchema.extend({
  subClassOf: z.array(TTIriRefSchema).prefault([]),
  code: z.string().optional(),
  im1Id: z.string().optional(),
  get matchedFrom(): z.ZodPrefault<z.ZodArray<typeof ConceptSchema>> {
    return z.array(ConceptSchema).prefault([]);
  },
  usage: z.number().optional(),
  codeId: z.string().optional(),
  alternativeCode: z.string().optional(),
  subsumed: z.boolean().default(false)
});

export type Concept = z.output<typeof ConceptSchema>;

export function isConcept(value: unknown): value is Concept {
  return ConceptSchema.safeParse(value).success;
}
